### PR TITLE
FileManager fixes

### PR DESCRIPTION
**Pull request by spevans**
- Use a common subfunction to read directory contents in
  contentsOfDirectory(atPath:) and subpathsOfDirectory(atPath:).

- Use readdir_r() instead of readdir().

- Call fileSystemRepresentation(withPath:) on the requested directory
  to match behaviour on Darwin.

- Replace the test case of a bad directory using "" with "/..." which
  now fails the precondition (to match Darwin).

### CODE REVIEW

**Comment by spevans** [comment]
@swift-ci please test

File: TestFoundation/TestFileManager.swift [hunk 1]
```diff
@@ -442,8 +442,8 @@ class TestFileManager : XCTestCase {
         }
         
         do {
-            let _ = try fm.contentsOfDirectory(atPath: "")
-            
+            // Check a bad path fails
```
Comment (millenomi) [inline TestFoundation/TestFileManager.swift]: Why the change here, for the record?

File: TestFoundation/TestFileManager.swift [hunk 2]
```diff
@@ -492,8 +492,8 @@ class TestFileManager : XCTestCase {
         }
         
         do {
-            let _ = try fm.subpathsOfDirectory(atPath: "")
-            
+            // Check a bad path fails
```
Comment (millenomi) [inline TestFoundation/TestFileManager.swift]: As above.

File: TestFoundation/TestFileManager.swift (hunk 1 above)
Comment (spevans) [reply TestFoundation/TestFileManager.swift]: Its for compatibility with Darwin's native Foundation. This call will throw an exception under native Foundation as the filename is passed through `fileSystemRepresentation(withPath:)` which treats `nil` and `""` as invalid paths. This PR modifies `contentsOfDirectory(atPath:)` and `subpathsOfDirectory(atPath:)` to now also pass the path through this function so `""` is now going to fail the `precondition(path != "", "Empty path argument")` check.

FYI, if you open the Xcode project `DarwinCompatibilityTests` in the `swift-corelibs-foundation` directory you can run all of the tests as unit tests against native Foundation to see what differences currently exist.

**Comment by spevans** [comment]
@swift-ci please test and merge
